src/main/java/io/aiven/elasticsearch/gcs/storage/EncryptionKeyMetadata.java

```diff
@@ -0,0 +1,93 @@
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+
+import io.aiven.elasticsearch.storage.security.EncryptionKeyProvider;
+
+import com.fasterxml.jackson.core.JsonEncoding;
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.core.JsonParseException;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonToken;
+import com.fasterxml.jackson.core.json.JsonWriteFeature;
+import com.google.cloud.Tuple;
+
+/** default */ class EncryptionKeyMetadata {
+
+    static JsonFactory jsonFactory;
+
+    static final int VERSION = 1;
+
+    static {
+        jsonFactory = new JsonFactory();
+        jsonFactory.configure(JsonWriteFeature.QUOTE_FIELD_NAMES.mappedFeature(), true);
+        jsonFactory.configure(JsonParser.Feature.ALLOW_COMMENTS, true);
+        jsonFactory.configure(JsonFactory.Feature.FAIL_ON_SYMBOL_HASH_OVERFLOW, false);
+        jsonFactory.configure(JsonGenerator.Feature.AUTO_CLOSE_JSON_CONTENT, false);
+        jsonFactory.configure(JsonParser.Feature.STRICT_DUPLICATE_DETECTION, true);
+    }
+
+    private final EncryptionKeyProvider encryptionKeyProvider;
+
+    public EncryptionKeyMetadata(final EncryptionKeyProvider encryptionKeyProvider) {
+        this.encryptionKeyProvider = encryptionKeyProvider;
+    }
+
+    public byte[] serialize(final SecretKey encryptionKey) throws IOException {
+        final var encryptedKey = encryptionKeyProvider.encryptKey(encryptionKey);
+
+        final var metadataBytes = new ByteArrayOutputStream();
+        try (final var jsonGenerator = jsonFactory.createGenerator(metadataBytes, JsonEncoding.UTF8)) {
+            jsonGenerator.writeStartObject();
+            jsonGenerator.writeStringField("key", Base64.getEncoder().encodeToString(encryptedKey));
+            jsonGenerator.writeNumberField("version", VERSION);
+            jsonGenerator.writeEndObject();
+        }
+        return metadataBytes.toByteArray();
+    }
+
+    public Tuple<SecretKey, Integer> deserialize(final byte[] metadata) throws IOException {
+        try {
+            final var jsonParser = jsonFactory.createParser(new String(metadata, StandardCharsets.UTF_8));
+            if (jsonParser.nextToken() != JsonToken.START_OBJECT) {
+                throw new IOException("Wrong JSON metadata");
+            }
+            jsonParser.nextToken();
+            final var encryptedKey = Base64.getDecoder().decode(jsonParser.nextTextValue());
+            jsonParser.nextToken();
+            final var version = jsonParser.nextIntValue(0);
+            if (version != VERSION) {
+                throw new IOException("Wrong JSON metadata");
+            }
+            if (jsonParser.nextToken() != JsonToken.END_OBJECT) {
+                throw new IOException("Wrong JSON metadata");
+            }
+            return Tuple.of(encryptionKeyProvider.decryptKey(encryptedKey), version);
```

Comment:
Sorry, this is not much better than binary encoding :) It will fail on pretty much any semantically invariant change in the JSON like swapping fields. What's the reason of not using high level Jackson part that encapsulate this and allows to read/write objects?

Reply:
true i will add databinding ... the problem is `jackson-core` is in the elastic main classpath, so it means that we need to keep the same version as in elastic ... will add comment into gradle.build

Reply:
I see, it's effectively useless for our situation. Well, in that case, it's fine to bring a JSON lib of choice (e.g. Gson)